feat(dm-integrity): add module for systemd-integritysetup

tasleson opened the pull request:
Adds module support for [systemd-integritysetup](https://github.com/systemd/systemd/pull/20902) which utilizes [dm-integrity](https://www.kernel.org/doc/html/latest/admin-guide/device-mapper/dm-integrity.html).

## Changes

## Checklist
- [x] I have tested it locally
- [x] I have reviewed and updated any documentation if relevant
- [ ] I am providing new code and test(s) for it

tasleson commented:
@dtardon Please let me know what I'm missing to get this PR merged, thanks!

johannbg commented:
@Conan-Kudo  why did you approve this PR and enable auto merge when the module  is half complete?

Conan-Kudo commented:
> @Conan-Kudo why did you approve this PR and enable auto merge when the module is half complete?

I did not know it was half-complete, sorry.

tasleson commented:
> > @Conan-Kudo why did you approve this PR and enable auto merge when the module is half complete?
> 
> I did not know it was half-complete, sorry.

@johannbg 

I understand that I'm missing something obvious to you, but if you could elaborate on why this PR is half complete I'll keep working on it.  From reviewing: https://github.com/dracutdevs/dracut/commit/3d4dea58f9821e58841d5c738b9935193c680181 and looking over https://github.com/dracutdevs/dracut/blob/master/docs/HACKING.md it's not obvious to me, thanks.

johannbg commented:
@tasleson your module should be constructed as systemd has designed it's support. So you should be familiar with how systemd implements things when writing systemd based modules not just our docs. 

For whatever reason you are missing an entire hostonly section ( which /etc/integritytab belongs in ) an hostsection which exist to allow administrators to include/overwrite existing type units for integritysetup ( you only included the vendor paths to the units ) along with the inclution of the library.  As I had already mentioned in the review,  take a look at the already existing systemd-veritysetup module to see how this module should look like and adapt accordingly for integritysetup.

Also minor nitpick fix your commit msg. 
you are introducing a new module so your commit message should reflect that ( feat(new_module_name): introduction msg ) so change it to something along the lines of... 
feat(systemd-integritysetup): introducing the systemd-integritysetup module

tasleson commented:
@johannbg 
> As I had already mentioned in the review ...

I'm not finding any previous review from you?  Thank you for the information you provided in this comment.  I believe I have added the missing parts and I've changed the commit msg. as suggested.

tasleson commented:
I re-did my testing and I'm running into some issues which I don't believe are related to these dracut changes.

* selinux issue: https://github.com/fedora-selinux/selinux-policy/issues/971
* Unable to mount dm-integrity volume in initramfs on first open: https://bugzilla.redhat.com/show_bug.cgi?id=2032527

johannbg commented:
> @johannbg
> 
> > As I had already mentioned in the review ...
> 
> I'm not finding any previous review from you? 

Oh the review is there, just in pending state in otherwords I did the review but have been disrupted before I submitted it, which also explains why you never responded to it ( I thought you were waiting for somekind of feedback from @dtardon ) I apologize for my mistake.

johannbg commented:
> * [](https://github.com/fedora-selinux/selinux-policy/issues/971)
> 
> * Unable to mount dm-integrity volume in initramfs on first open: https://bugzilla.redhat.com/show_bug.cgi?id=2032 
I suspect a bug in the generator. unit not created correctly, runtime directory missing, something along those lines

tasleson commented:
@johannbg 
>       I suspect a bug in the generator. unit not created correctly, runtime directory missing, something along those lines

The device mapper integrity target is up when I fall into the emergency shell which is all we expect out of the systemd generator and setup.  blkid reports all the correct information for the integrity device.  systemd does a fsck on the integrity device which is good, we just can't mount it.  If is close/open the integrity target we can mount the FS.

I'm continuing investigation, but I don't have reason to believe it's the systemd-integritysetup or systemd-integritygenerator at the moment.

tasleson commented:
@johannbg 

> I'm continuing investigation, but I don't have reason to believe it's the systemd-integritysetup or systemd-integritygenerator at the moment.

Latest data is showing that dm-integrity is inducing corruption in my fedora rawhide testing, ref. https://bugzilla.redhat.com/show_bug.cgi?id=2032527#c4

johannbg commented:
@tasleson to me this looks like a race condition. 
`
Dec 14 09:19:42 dracut systemd[1]: Mounting /sysroot...
`
`
Dec 14 09:19:42 dracut mount[366]: mount: /sysroot: wrong fs type, bad option, bad superblock on /dev/mapper/integrityroot, missing codepage or helper program, or other error.
`
Something does not seem to be ready to be mounted at this point ( or is already mounted due to the fsck that seems to be running prior as in this `dracut systemd-fsck[365]: /dev/mapper/integrityroot`). 

Followed by this working
`
[root@dracut ~]# integritysetup close integrityroot
`
`
[root@dracut ~]# integritysetup open /dev/sdc1 integrityroot
`
`
[root@dracut ~]# mount /dev/mapper/integrityroot /mnt
`
Would lead me to look at/double check the ordering dependency ( Before, After etc ) in the units, the systemd-integritysetup-generator creates.

tasleson commented:
@johannbg 
> @tasleson to me this looks like a race condition.

I would agree, but once I discovered that you can't even create a xfs FS on a freshly formatted integrity device, leads us to something is really broken in the device mapper integrity target (or in the kernel itself) on fedora rawhide.  Previous testing with proposed dracut changes on F34 worked.  I selected rawhide for these latest changes as it includes the upstream systemd integrity changes.

I chatted with the device mapper folks yesterday, they are taking a look.

tasleson commented:
Fully tested on  F35 with a locally updated systemd packages built to include upstream `systemd-integrity*` files.  The issue I was running into is indeed a dm-integrity bug.  Fix included in 5.16rc6.

tasleson commented:
@Conan-Kudo This PR is ready now if you could take a review so we could get it merged, thanks!
